### PR TITLE
Add the Runtime Cache Design Docs

### LINKED ISSUE #679: Design the Runtime Cache
### Summary
In order to efficiently and effectively cache intermediate results at runtime, we need to ensure that we have an appropriate underlying structure. This task deals with coming up with a design that satisfies the resource constraints and functional requirements placed on the runtime cache. 

### Value
We will have a design for the runtime cache that satisfies the requirements placed on it. 

### Specification

- [ ] Determine the optimal underlying data structure for storing a key -> value mapping that will allow us to control memory usage.
- [ ] Optimally, we want to be able to put a maximum bound on the size of the cache (look into the GraalVM Resource Limits API) in memory.
- [ ] We want to record LFU data to aid eviction of entries when the size is large.
- [ ] The eviction strategy should also account for whether the value is currently being _visualised_, as if it isn't being visualised it need not be cached. 
- [ ] Document the specification for the cache in `doc/runtime/specification/caching.md`.
- [ ] Document any planned implementation details for the cache in `doc/runtime/implementation/caching.md`.
- [ ] Document any reasoning behind the specification for the cache in `doc/runtime/design/caching.md`.

### Acceptance Criteria & Test Cases

- We have a concrete design for the runtime cache.
- That design has been documented in the design documentation.

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #679

Changelog:

- doc: add runtime cache design

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/docs/style-guide/scala.md) and [Java](https://github.com/luna/enso/blob/master/docs/style-guide/java.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: docs/runtime/caching.md [hunk 1]
```diff
@@ -60,15 +71,37 @@ foo a b =
...
+Initial version caches only the right-hand side expressions of the
+bindings. This way, we can guarantee that between the executions, the cache
+stays within the memory limit without implementing [Memory
+Management](#memorymanagement), the eviction strategy based on the overall size
+of the cache.
+
+### Further Development of Cache Candidates
+The next step is caching intermediate expressions. With this approach, we need a
+way to control the overall size of the cache because we can no longer guarantee
+the safe memory limits between the two program executions. Consider the example:
```
Comment (iamrecursion) [inline docs/runtime/caching.md:94]:
```suggestion
The next step for the cache is to expand the portions of the introspected scope
that we cache. In general, this means the caching of intermediate expressions.

However, once we do this, we can no longer guarantee that we do not push the
JVM out of memory between two program executions. This is best demonstrated
by example. 
```

File: docs/runtime/caching.md [hunk 2]
```diff
@@ -60,15 +71,37 @@ foo a b =
...
+
+```
+a = (computeHugeObject b).size
+```
+
+Here we compute a considerably large object and discard it, saving only a single
+field - its size. Hence if we want to cache the intermediate result of discarded
+`computeHubeObject b` expression, we need to implement the mechanism of tracking
+the overall size of the cache. See [Memory Management](#memorymanagement)
+section.
```
Comment (iamrecursion) [inline docs/runtime/caching.md:104]:
```suggestion
Here we compute a value that takes up a significant amount of memory, but from it we
only compute a small derived value (its size). Hence, if we want to cache the intermediate
result of the discarded `computeHugeObject b` expression, we need some way of tracking
the sizes of individual cache entries. 
```

File: docs/runtime/caching.md [hunk 3]
```diff
@@ -344,5 +377,100 @@ changed:
 2. If `k` is a dynamic symbol, all expressions that depend on _any instance_ of
    the dynamic symbol are evicted from the cache.
 
-Expressions that have been evicted from the cache subsequently have to be
-recomputed by the runtime.
+Expressions that have been evicted from the cache subsequently have to be recomputed by the runtime.
```
Comment (iamrecursion) [inline docs/runtime/caching.md:380]: This overflows.

File: docs/runtime/caching.md [hunk 4]
```diff
@@ -344,5 +377,100 @@ changed:
...
+
+Cache eviction takes into account the following aspects:
+
+Visualization. In the first place, we should care about nodes that have
```
Comment (iamrecursion) [inline docs/runtime/caching.md:384]: This should be a bulleted list using the standard format with bolded headers.

File: docs/runtime/caching.md [hunk 5]
```diff
@@ -344,5 +377,100 @@ changed:
...
+visualization attached in the IDE. Priority. The runtime can assign a score to
+a node, meaning how valuable this node is. Less valuable nodes should be evicted
+first. Computation time. The runtime can calculate the time that node took to
+compute. Less computationally intensive nodes should be evicted first. Memory
+limit. The cache should not exceed the specified memory limit.
+
+## Cache Backend
+The requirements for eviction strategies and memory management should guide the
```
Comment (iamrecursion) [inline docs/runtime/caching.md:392]: This doesn't say anything useful.

File: docs/runtime/caching.md [hunk 6]
```diff
@@ -344,5 +377,100 @@ changed:
...
+Expressions that have been evicted from the cache subsequently have to be recomputed by the runtime.
+
+Cache eviction takes into account the following aspects:
+
+Visualization. In the first place, we should care about nodes that have
+visualization attached in the IDE. Priority. The runtime can assign a score to
+a node, meaning how valuable this node is. Less valuable nodes should be evicted
+first. Computation time. The runtime can calculate the time that node took to
+compute. Less computationally intensive nodes should be evicted first. Memory
+limit. The cache should not exceed the specified memory limit.
```
Comment (iamrecursion) [inline docs/runtime/caching.md:389]: This doesn't actually specify the eviction strategy. This document should contain a rigorous algorithmic spec of how we do eviction.

File: docs/runtime/caching.md [hunk 7]
```diff
@@ -344,5 +377,100 @@ changed:
...
+
+## Cache Backend
+The requirements for eviction strategies and memory management should guide the
+exact implementation of the cache.
+
+### Initial Implementation of Cache Backend
+The cache consists of key-value storage and an eviction function. Key-value
```
Comment (iamrecursion) [inline docs/runtime/caching.md:396]: This is very light on detail. You need to describe the _what_ and the _why_ of the design and implementation of this cache. This section should also mention the details, such as the use of soft references as a fallback strategy.

File: docs/runtime/caching.md [hunk 8]
```diff
@@ -344,5 +377,100 @@ changed:
...
+storage can be as simple as a Hash Map with values wrapped into the Soft
+References as a safety measure. Eviction function purges invalidated expressions
+based on the [Cache Eviction Strategy](#cacheevictionstrategy).
+
+### Further Development of Cache Backend
+The backend implementation should be updated according to the changing
```
Comment (iamrecursion) [inline docs/runtime/caching.md:402]: First sentence here is useless. This section should instead be about how we plan on extending it in the future, not just examples. Furthermore it should have a standard `> The actionables for this section are` block, just as other documents do.

File: docs/runtime/caching.md [hunk 9]
```diff
@@ -344,5 +377,100 @@ changed:
...
+requirements of the runtime cache.For example, to extend the eviction strategy,
+we can add an asynchronous scoring task that computes some properties of stored
+objects (e.g., the size of the object). Those properties can be used in the
+eviction strategy as optional clues, improving the hit ratio.
+
+## Memory Management
```
Comment (iamrecursion) [inline docs/runtime/caching.md:408]: This isn't a good title. It should be "Memory Bounded Caches" or similar.

File: docs/runtime/caching.md [hunk 10]
```diff
@@ -344,5 +377,100 @@ changed:
...
+Memory management refers to a way of controlling the size of the cache, avoiding
+the Out Of Memory errors.
+
+The methods below can be divided into two approaches.
+
+1. Limiting overall JVM memory and relying on garbage collection
```
Comment (iamrecursion) [inline docs/runtime/caching.md:414]:
```suggestion
1. Limiting the overall JVM memory and relying on garbage collection.
```

File: docs/runtime/caching.md [hunk 11]
```diff
@@ -344,5 +377,100 @@ changed:
...
+2. Calculating the object's size and using it in the eviction policy
```
Comment (iamrecursion) [inline docs/runtime/caching.md:415]:
```suggestion
2. Calculating the object's size and using it in the eviction strategy.
```

File: docs/runtime/caching.md [hunk 12]
```diff
@@ -344,5 +377,100 @@ changed:
...
+
+In general, to control the memory size on JVM, anything besides limiting the
+total amount involves some tradeoff.
```
Comment (iamrecursion) [inline docs/runtime/caching.md:418]:
```suggestion
total amount involves some tradeoffs.
```

File: docs/runtime/caching.md [hunk 13]
```diff
@@ -344,5 +377,100 @@ changed:
...
+
+### Soft References
+Soft References is a way to mark the cache entries available for garbage
```
Comment (iamrecursion) [inline docs/runtime/caching.md:421]: This section should be broken up into multiple paragraphs.

File: docs/runtime/caching.md [hunk 14]
```diff
@@ -344,5 +377,100 @@ changed:
...
+JVM, running in a thread with their own heap, memory limit, and garbage
+collection. The problem is that _isolates_ can't share objects. And even if we
+move the cache to the separate _isolate_, that would require creating a
+mechanism of sharing objects based on pointers, which requires implementing
+serialization.  On the other hand, serialization itself can provide the size of
+the object, which is enough to implement the eviction policy, even without
+running the _isolates_.
+
+### Serialization
+One approach to get the size of the value stored in the cache is
```
Comment (iamrecursion) [inline docs/runtime/caching.md:435]: This section and its siblings are very light on detail. If you did benchmarking of these overheads please include numbers.

File: docs/runtime/caching.md [hunk 15]
```diff
@@ -344,5 +377,100 @@ changed:
...
+approach requires running JVM with _javaagent_ attached, which can complicate
+the deployment and have indirect performance penalties.
+
+### Manual Memory Management
+This method implies tracking the size of the values by hand. It can be done for
+the values of Enso language, knowing the size of its primitive types. For
+interacting with other languages, Enso can provide an interface that developers
+should implement to have a better experience with the cache.
+
+### Comparison of Memory Management Approaches
```
Comment (iamrecursion) [inline docs/runtime/caching.md:450]: This really should include actual numbers.

File: docs/runtime/caching.md [hunk 16]
```diff
@@ -60,15 +71,37 @@ foo a b =
     a.quux d
 ```
 
-The cache is active for the _currently visible scope_ in Enso Studio, so when a
-user enters the function `foo`, the cache stores the intermediate results in
-this function (in this case `c` and `d`), as well as the inputs to the function
-(in this case `a`, and `b`).
+The cache is active for the currently visible scope in Enso Studio, so when a
```
[the remark this answers is not in the document]
Comment (4e6) [reply docs/runtime/caching.md:74]: Can't imagine how I missed that. I feel so absent-minded lately :facepalm:

**Pull request merged**